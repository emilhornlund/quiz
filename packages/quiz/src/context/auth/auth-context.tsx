import { createContext } from 'react'

import { Client, Player } from '../../models'

/**
 * Represents the structure of the authentication context.
 *
 * @property token - The authentication token (optional).
 * @property client - The client information (optional).
 * @property player - The player information (optional).
 * @property setToken - Function to update the authentication token.
 * @property setClient - Function to update the client information.
 * @property setPlayer - Function to update the player information.
 */
export type AuthContextType = {
  token?: string
  client?: Client
  player?: Player
  setToken: (token: string) => void
  setClient: (client: Client) => void
  setPlayer: (player: Player) => void
}

/**
 * A React context providing authentication-related state and functions.
 *
 * The default context value initializes `setToken`, `setClient`, and `setPlayer`
 * as no-op functions.
 */
export const AuthContext = createContext<AuthContextType>({
  setToken: () => undefined,
  setClient: () => undefined,
  setPlayer: () => undefined,
})
