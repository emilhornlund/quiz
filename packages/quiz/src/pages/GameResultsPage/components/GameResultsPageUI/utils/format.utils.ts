/**
 * Converts a duration in seconds to a rounded, human-readable format.
 *
 * @param seconds - The total duration in seconds.
 * @returns A formatted string, e.g., "2 hours, 5 minutes".
 */
export function formatRoundedDuration(seconds: number): string {
  const totalMinutes = Math.round(seconds / 60)
  const hours = Math.floor(totalMinutes / 60)
  const minutes = totalMinutes % 60

  const parts: string[] = []

  if (hours > 0) parts.push(`${hours} hour${hours !== 1 ? 's' : ''}`)
  if (minutes > 0 || parts.length === 0) {
    parts.push(`${minutes} minute${minutes !== 1 ? 's' : ''}`)
  }

  return parts.join(', ')
}

/**
 * Converts a duration in milliseconds to a rounded string in seconds.
 *
 * @param milliseconds - The duration in milliseconds.
 * @returns A string like "2.5s" or "3s", depending on rounding.
 */
export function formatRoundedSeconds(milliseconds: number): string {
  const seconds = milliseconds / 1000
  const rounded = Math.round(seconds * 10) / 10
  return rounded % 1 === 0 ? `${rounded.toFixed(0)}s` : `${rounded.toFixed(1)}s`
}
