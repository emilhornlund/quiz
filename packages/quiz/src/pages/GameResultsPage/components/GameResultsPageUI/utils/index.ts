export * from './format.util.ts'
export * from './metric.utils.ts'
