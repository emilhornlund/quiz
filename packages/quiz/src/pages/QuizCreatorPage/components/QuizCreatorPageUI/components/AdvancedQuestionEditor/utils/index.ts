export * from './parse-questions-json'
