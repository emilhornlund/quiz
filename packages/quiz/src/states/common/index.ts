export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export type { QuestionAnswerPickerProps } from './QuestionAnswerPicker'
export type { QuestionMediaProps } from './QuestionMedia'
export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'

export { default as GamePage } from './GamePage'
export { default as NonInteractiveInfoBox } from './NonInteractiveInfoBox'
export { default as PointsBehindIndicator } from './PointsBehindIndicator'
export { default as QuestionAnswerPicker } from './QuestionAnswerPicker'
export { default as QuestionMedia } from './QuestionMedia'
export { default as QuestionTextPreview } from './QuestionTextPreview'
export { default as QuestionTypePointsBar } from './QuestionTypePointsBar'
