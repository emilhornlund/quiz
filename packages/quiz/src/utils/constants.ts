export const LEGACY_HOST = 'quiz.emilhornlund.com'
export const REDIRECT_TARGET_HOST = 'klurigo.com'
