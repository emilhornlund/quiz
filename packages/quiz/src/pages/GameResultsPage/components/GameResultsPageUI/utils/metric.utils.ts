import { IconDefinition } from '@fortawesome/fontawesome-common-types'
import {
  faCheckCircle,
  faFire,
  faMinusCircle,
  faStar,
  faStopwatch,
  faTimesCircle,
} from '@fortawesome/free-solid-svg-icons'
import {
  GameMode,
  GameResultClassicModePlayerMetricDto,
  GameResultClassicModeQuestionMetricDto,
  GameResultDto,
  GameResultZeroToOneHundredModePlayerMetricDto,
  GameResultZeroToOneHundredModeQuestionMetricDto,
} from '@quiz/common'

import colors from '../../../../../styles/colors.module.scss'

import { formatRoundedSeconds } from './format.utils.ts'

/**
 * Calculates the percentage of correct answers for a given metric.
 *
 * @param metric - The metric containing counts of correct, incorrect, and unanswered answers.
 * @returns A percentage (0–100) rounded up to the nearest integer.
 */
export function getCorrectPercentage(
  metric:
    | GameResultClassicModePlayerMetricDto
    | GameResultClassicModeQuestionMetricDto,
): number {
  const { correct, incorrect, unanswered } = metric
  const total = correct + incorrect + unanswered
  return total > 0 ? Math.ceil((correct / total) * 100) : 0
}

/**
 * Converts the average precision (0–1) into a percentage format.
 *
 * @param metric - The metric containing the average precision.
 * @returns A percentage (0–100) rounded up to the nearest integer.
 */
export function getAveragePrecision(
  metric:
    | GameResultZeroToOneHundredModePlayerMetricDto
    | GameResultZeroToOneHundredModeQuestionMetricDto,
): number {
  const { averagePrecision } = metric
  return Math.ceil(averagePrecision * 100)
}

/**
 * Returns a descriptive message based on the overall correctness percentage.
 *
 * @param percentage - The overall correctness percentage (0–100).
 * @returns A message describing the difficulty level of the quiz.
 */
export function getQuizDifficultyMessage(percentage: number): string {
  if (percentage < 20) {
    return 'This one was a brain-buster. Tougher than most could handle!'
  } else if (percentage < 40) {
    return 'Challenging quiz! Only a few managed to crack these questions.'
  } else if (percentage < 60) {
    return 'A balanced battle — tricky, but not impossible.'
  } else if (percentage < 80) {
    return 'Well played! Most players handled the challenge with confidence.'
  } else if (percentage < 95) {
    return 'Nice and smooth — players breezed through most questions!'
  } else {
    return "Too easy! Next time, let's turn up the difficulty."
  }
}

/**
 * Represents details about a metric to be displayed in the results UI.
 */
type MetricDetails = {
  title: string
  value: string | number
  icon?: IconDefinition
  iconColor?: string
}

/**
 * Builds an array of metric details for a player's performance, customized by game mode.
 *
 * @param mode - The game mode (e.g., Classic, ZeroToOneHundred).
 * @param metric - The player metric object with performance data.
 * @returns An array of metric details for display in the player section.
 */
export function buildPlayerSectionMetricDetails(
  mode: GameMode,
  metric: GameResultDto['playerMetrics'][0],
): MetricDetails[] {
  const common: MetricDetails[] = [
    {
      title: 'Unanswered',
      value: metric.unanswered,
      icon: faMinusCircle,
      iconColor: colors.gray2,
    },
    {
      title: 'Average response time',
      value: formatRoundedSeconds(metric.averageResponseTime),
      icon: faStopwatch,
      iconColor: colors.turquoise2,
    },
    {
      title: 'Longest correct streak',
      value: metric.longestCorrectStreak,
      icon: faFire,
      iconColor: colors.orange2,
    },
    {
      title: 'Score',
      value: metric.score,
      icon: faStar,
      iconColor: colors.yellow2,
    },
  ]
  if (mode === GameMode.Classic) {
    const classicModePlayerMetric =
      metric as GameResultClassicModePlayerMetricDto
    return [
      {
        title: 'Correct',
        value: classicModePlayerMetric.correct,
        icon: faCheckCircle,
        iconColor: colors.green2,
      },
      {
        title: 'Incorrect',
        value: classicModePlayerMetric.incorrect,
        icon: faTimesCircle,
        iconColor: colors.red2,
      },
      ...common,
    ]
  }
  return common
}

/**
 * Builds an array of metric details for a question's performance, customized by game mode.
 *
 * @param mode - The game mode (e.g., Classic, ZeroToOneHundred).
 * @param metric - The question metric object with performance data.
 * @returns An array of metric details for display in the question section.
 */
export function buildQuestionSectionMetricDetails(
  mode: GameMode,
  metric: GameResultDto['questionMetrics'][0],
): MetricDetails[] {
  const common: MetricDetails[] = [
    {
      title: 'Unanswered',
      value: metric.unanswered,
      icon: faMinusCircle,
      iconColor: colors.gray2,
    },
    {
      title: 'Average response time',
      value: formatRoundedSeconds(metric.averageResponseTime),
      icon: faStopwatch,
      iconColor: colors.turquoise2,
    },
  ]
  if (mode === GameMode.Classic) {
    const classicModePlayerMetric =
      metric as GameResultClassicModeQuestionMetricDto
    return [
      {
        title: 'Correct',
        value: classicModePlayerMetric.correct,
        icon: faCheckCircle,
        iconColor: colors.green2,
      },
      {
        title: 'Incorrect',
        value: classicModePlayerMetric.incorrect,
        icon: faTimesCircle,
        iconColor: colors.red2,
      },
      ...common,
    ]
  }
  return common
}
