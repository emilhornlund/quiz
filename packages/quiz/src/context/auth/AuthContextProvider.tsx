import React, {
  FC,
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react'
import { useNavigate } from 'react-router-dom'

import { useQuizServiceClient } from '../../api/use-quiz-service-client.tsx'
import { AuthState } from '../../models'
import { AUTH_LOCAL_STORAGE_KEY } from '../../utils/constants.ts'

import { AuthContext, AuthContextType } from './auth-context.tsx'

/**
 * Parses a JSON string into an object of type `T`, or returns `undefined` if the string is falsy.
 *
 * @template T - The expected type of the parsed object.
 * @param jsonString - The JSON string to parse.
 * @returns The parsed object of type `T` or `undefined` if the input is invalid.
 */
const parseJSONString = <T extends AuthState>(
  jsonString?: string | null,
): T | undefined => {
  if (!jsonString) {
    return undefined
  }
  return JSON.parse(jsonString) as T
}

/**
 * Props for the `AuthContextProvider` component.
 *
 * @property children - The child components to be wrapped by the provider.
 */
export interface AuthContextProviderProps {
  children: ReactNode | ReactNode[]
}

/**
 * A context provider for managing authentication-related state.
 *
 * It synchronizes the context state with local storage and provides functions to update
 * and persist authentication data.
 *
 * @param children - The child components to be wrapped by the provider.
 * @returns A React component wrapping its children with the `AuthContext` provider.
 */
const AuthContextProvider: FC<AuthContextProviderProps> = ({ children }) => {
  const { revoke } = useQuizServiceClient()

  const navigate = useNavigate()

  const [auth, setAuth] = useState<AuthState>()

  /**
   * Determines whether the user is currently logged in by verifying
   * the presence of both access and refresh tokens in state.
   */
  const isLoggedIn = useMemo(
    () => !!(auth?.accessToken && auth?.refreshToken),
    [auth],
  )

  useEffect(() => {
    setAuth(
      parseJSONString<AuthState>(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)),
    )
  }, [])

  /**
   * Updates the auth information in the state and persists it to local storage.
   *
   * @param newAuth - The new auth object to set.
   */
  const handleSetAuth = (newAuth?: AuthState) => {
    setAuth(newAuth)
    if (newAuth) {
      localStorage.setItem(AUTH_LOCAL_STORAGE_KEY, JSON.stringify(newAuth))
    } else {
      localStorage.removeItem(AUTH_LOCAL_STORAGE_KEY)
    }
  }

  /**
   * Logs out the current user by revoking their authentication token,
   * clearing the stored authentication state, and navigating to the home page.
   *
   * If an access or refresh token exists, it will be sent to the server for revocation.
   */
  const handleLogout = useCallback(() => {
    const token = auth?.accessToken || auth?.refreshToken
    if (token) {
      revoke({ token }).then(() => {
        handleSetAuth(undefined)
        navigate('/')
      })
    }
  }, [auth, revoke, navigate])

  /**
   * Memoized value for the `AuthContext`, containing the current authentication state
   * and update functions.
   */
  const value = useMemo<AuthContextType>(
    () => ({
      accessToken: auth?.accessToken,
      refreshToken: auth?.refreshToken,
      isLoggedIn,
      setAuth: handleSetAuth,
      logout: handleLogout,
    }),
    [auth, isLoggedIn, handleLogout],
  )

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}

export default AuthContextProvider
