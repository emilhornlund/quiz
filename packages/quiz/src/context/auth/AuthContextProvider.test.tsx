import { TokenScope, TokenType } from '@quiz/common'
import { act, cleanup, render, waitFor } from '@testing-library/react'
import React, { useContext, useEffect } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

import { AUTH_LOCAL_STORAGE_KEY } from '../../utils/constants'

import { AuthContext } from './auth-context'
import AuthContextProvider from './AuthContextProvider'

// --- Mocks ---
const mockRevoke = vi.fn().mockResolvedValue({})
vi.mock('../../api/use-quiz-service-client.tsx', () => ({
  useQuizServiceClient: () => ({ revoke: mockRevoke }),
}))

const mockNavigate = vi.fn()
vi.mock('react-router-dom', async (importActual) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const actual = await importActual<any>()
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  }
})

const mockJwtDecode = vi.fn()
// Keep the payload shape simple; we just need to see it flow through.
vi.mock('jwt-decode', () => ({
  jwtDecode: (token: string) =>
    mockJwtDecode(token) ??
    ({
      sub: `sub:${token}`,
      exp: 123,
      authorities: ['ROLE_USER'],
      // optional fields for game scope:
      gameId: token.includes('game') ? 'game-123' : undefined,
      participantType: token.includes('game') ? 'Player' : undefined,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any),
}))

// --- Test helpers ---
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function renderWithProvider(capture: (ctx: any) => void, ui?: React.ReactNode) {
  const Probe: React.FC = () => {
    const ctx = useContext(AuthContext)
    useEffect(() => {
      capture(ctx)
    }, [ctx])
    return ui ?? null
  }
  return render(
    <BrowserRouter>
      <AuthContextProvider>
        <Probe />
      </AuthContextProvider>
    </BrowserRouter>,
  )
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function setLocalStorageAuth(state: any) {
  localStorage.setItem(AUTH_LOCAL_STORAGE_KEY, JSON.stringify(state))
}

// --- Lifecycle ---
beforeEach(() => {
  vi.useRealTimers()
  localStorage.clear()
  mockRevoke.mockClear()
  mockNavigate.mockClear()
  mockJwtDecode.mockClear()
})

afterEach(() => {
  cleanup()
})

// --- Tests ---
describe('AuthContextProvider', () => {
  it('initializes from localStorage on first render (no flicker)', async () => {
    // Pre-populate localStorage with a USER token so the first render is authenticated
    setLocalStorageAuth({
      USER: {
        [TokenType.Access]: { token: 'user.access', exp: 1 },
        [TokenType.Refresh]: { token: 'user.refresh', exp: 2 },
      },
      GAME: undefined,
    })

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    await waitFor(() => {
      expect(ctx).toBeTruthy()
      expect(ctx.isUserAuthenticated).toBe(true)
      expect(ctx.isGameAuthenticated).toBe(false)
      expect(ctx.user?.ACCESS.token).toBe('user.access')
    })

    expect(container).toMatchSnapshot()
  })

  it('starts unauthenticated when localStorage is empty', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    await waitFor(() => {
      expect(ctx.isUserAuthenticated).toBe(false)
      expect(ctx.isGameAuthenticated).toBe(false)
      expect(ctx.user).toBeUndefined()
      expect(ctx.game).toBeUndefined()
    })

    expect(container).toMatchSnapshot()
  })

  it('setTokenPair(User) decodes tokens and persists to localStorage', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    await act(async () => {
      ctx.setTokenPair(TokenScope.User, 'user.access.jwt', 'user.refresh.jwt')
    })

    await waitFor(() => {
      // context updated
      expect(ctx.isUserAuthenticated).toBe(true)
      expect(ctx.user?.ACCESS.token).toBe('user.access.jwt')
      expect(ctx.user?.REFRESH.token).toBe('user.refresh.jwt')
      // persisted
      const parsed = JSON.parse(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)!)
      expect(parsed.USER.ACCESS.token).toBe('user.access.jwt')
      expect(parsed.USER.REFRESH.token).toBe('user.refresh.jwt')
    })

    expect(container).toMatchSnapshot()
  })

  it('setTokenPair(Game) decodes game fields and persists to localStorage', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    await act(async () => {
      ctx.setTokenPair(TokenScope.Game, 'game.access.jwt', 'game.refresh.jwt')
    })

    await waitFor(() => {
      expect(ctx.isGameAuthenticated).toBe(true)
      expect(ctx.game?.ACCESS.token).toBe('game.access.jwt')
      expect(ctx.game?.REFRESH.token).toBe('game.refresh.jwt')
      // game extras from mocked decoder
      expect(ctx.game?.ACCESS.gameId).toBe('game-123')
      expect(ctx.game?.ACCESS.participantType).toBe('Player')
      // persisted
      const parsed = JSON.parse(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)!)
      expect(parsed.GAME.ACCESS.token).toBe('game.access.jwt')
    })

    expect(container).toMatchSnapshot()
  })

  it('clears per-scope auth with revokeUser/revokeGame (calls API, navigates, updates storage)', async () => {
    // seed both scopes
    setLocalStorageAuth({
      USER: {
        [TokenType.Access]: { token: 'u.acc', exp: 1 },
        [TokenType.Refresh]: { token: 'u.ref', exp: 2 },
      },
      GAME: {
        [TokenType.Access]: { token: 'g.acc', exp: 3 },
        [TokenType.Refresh]: { token: 'g.ref', exp: 4 },
      },
    })

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    // revoke user first
    await act(async () => {
      await ctx.revokeUser()
    })

    await waitFor(() => {
      expect(mockRevoke).toHaveBeenCalledWith({ token: 'u.acc' }) // access is preferred
      expect(mockNavigate).toHaveBeenCalledWith('/')
      expect(ctx.user).toBeUndefined()
      // game still present
      expect(ctx.game).toBeTruthy()
    })

    // revoke game next
    await act(async () => {
      await ctx.revokeGame()
    })

    await waitFor(() => {
      expect(mockRevoke).toHaveBeenCalledWith({ token: 'g.acc' })
      expect(ctx.game).toBeUndefined()
      // storage cleared entirely when both scopes are empty
      expect(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)).toBeNull()
    })

    expect(container).toMatchSnapshot()
  })

  it('removes storage entry when both scopes are cleared', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    await act(async () => {
      ctx.setTokenPair(TokenScope.User, 'user.acc', 'user.ref')
      ctx.setTokenPair(TokenScope.Game, 'game.acc', 'game.ref')
    })
    expect(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)).toBeTruthy()

    await act(async () => {
      await ctx.revokeUser()
    })
    await act(async () => {
      await ctx.revokeGame()
    })

    await waitFor(() => {
      expect(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)).toBeNull()
    })

    expect(container).toMatchSnapshot()
  })

  it('removes storage when the last remaining scope is cleared', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let ctx: any
    const { container } = renderWithProvider((c) => (ctx = c))

    await act(async () => {
      ctx.setTokenPair(TokenScope.User, 'user.acc', 'user.ref')
    })
    expect(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)).toBeTruthy()

    await act(async () => {
      await ctx.revokeUser()
    })

    await waitFor(() => {
      expect(localStorage.getItem(AUTH_LOCAL_STORAGE_KEY)).toBeNull()
    })

    expect(container).toMatchSnapshot()
  })
})
