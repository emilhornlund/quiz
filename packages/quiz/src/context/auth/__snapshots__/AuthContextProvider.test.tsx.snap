// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthContextProvider > clears per-scope auth with revokeUser/revokeGame (calls API, navigates, updates storage) 1`] = `<div />`;

exports[`AuthContextProvider > initializes from localStorage on first render (no flicker) 1`] = `<div />`;

exports[`AuthContextProvider > removes storage entry when both scopes are cleared 1`] = `<div />`;

exports[`AuthContextProvider > removes storage when the last remaining scope is cleared 1`] = `<div />`;

exports[`AuthContextProvider > revokeUser swallows errors and still clears & navigates 1`] = `<div />`;

exports[`AuthContextProvider > setTokenPair(Game) decodes game fields and persists to localStorage 1`] = `<div />`;

exports[`AuthContextProvider > setTokenPair(User) decodes tokens and persists to localStorage 1`] = `<div />`;

exports[`AuthContextProvider > starts unauthenticated when localStorage is empty 1`] = `<div />`;
